handlers/api.go

```diff
@@ -839,6 +848,177 @@ func getSyncCommitteeFor(validators []uint64, period uint64) ([]interface{}, err
+		}
+		uniquePeriods[utils.SyncPeriodOfEpoch(firstSyncEpoch)] = true
+
+		// exit epoch (if any)
+		if v.ExitEpoch != noExitEpoch && v.ExitEpoch > firstSyncEpoch {
+			uniquePeriods[utils.SyncPeriodOfEpoch(v.ExitEpoch)] = true
+		}
+	}
+
+	// transform map to slice; this will be used to query sync_committees_count_per_validator
+	timeStampSlice := make([]uint64, 0, len(uniquePeriods))
+	for ts := range uniquePeriods {
+		timeStampSlice = append(timeStampSlice, ts)
+	}
+
+	// get aggregated count for all relevant committees from sync_committees_count_per_validator
+	var countStatistics []struct {
+		Period     uint64  `db:"period"`
+		CountSoFar float64 `db:"count_so_far"`
+	}
+
+	query, args, errs := sqlx.In(`SELECT period, count_so_far FROM sync_committees_count_per_validator WHERE period IN (?) ORDER BY period ASC`, timeStampSlice)
+	if errs != nil {
+		return nil, errs
+	}
+	err = db.ReaderDb.Select(&countStatistics, db.ReaderDb.Rebind(query), args...)
+	if err != nil {
+		return nil, err
+	}
+
+	// transform query result to map for easy access
+	periodInfoMap := make(map[uint64]float64)
+	for _, pl := range countStatistics {
+		periodInfoMap[pl.Period] = pl.CountSoFar
+	}
+
+	// calculate expected committies for every single validator and aggregate them
+	expectedCommitties := 0.0
+	for _, vi := range validatorsEpochInfo {
+		if _, found := periodInfoMap[utils.SyncPeriodOfEpoch(vi.ActivationEpoch)]; !found {
+			return nil, fmt.Errorf("required period not found")
+		}
+
+		lastEpoch := vi.ExitEpoch
+		if vi.ExitEpoch == noExitEpoch {
+			lastEpoch = epoch
+		}
+
+		if _, found := periodInfoMap[utils.SyncPeriodOfEpoch(lastEpoch)]; !found {
+			return nil, fmt.Errorf("required period not found")
+		}
+
+		expectedCommitties += periodInfoMap[utils.SyncPeriodOfEpoch(lastEpoch)] - periodInfoMap[utils.SyncPeriodOfEpoch(vi.ActivationEpoch)]
+	}
+
+	// transform committees to slots
+	r.ExpectedSlots = uint64(expectedCommitties * float64(utils.Config.Chain.Config.EpochsPerSyncCommitteePeriod*utils.Config.Chain.Config.SlotsPerEpoch))
+
+	////
+	// calculate ParticipatedSlots and MissedSlots
+
+	// collect aggregated sync committee stats from validator_stats table for all validators
+	var syncStats struct {
+		Participated int64 `db:"participated"`
+		Missed       int64 `db:"missed"`
+	}
+	query, args, err = sqlx.In(`SELECT COALESCE(SUM(participated_sync), 0) AS participated, COALESCE(SUM(missed_sync), 0) AS missed FROM validator_stats WHERE validatorindex IN (?)`, validators)
+	if err != nil {
+		return nil, err
+	}
+	err = db.ReaderDb.Get(&syncStats, db.ReaderDb.Rebind(query), args...)
+	if err != nil {
+		return nil, err
+	}
+
+	r.ParticipatedSlots = uint64(syncStats.Participated)
+	r.MissedSlots = uint64(syncStats.Missed)
+
+	// validator_stats is updated only once a day, everything missing has to be collected from bigtable (which is slower than validator_stats)
+	// check when the last update to validator_stats was
+	var lastExportedDay uint64
```

Comment:
Can we move the loading of fresh stats from bigtable to it's own function? This function is getting me too big

Reply:
Sure can do, changed.